File: bzk/bzkflow/src/main/java/net/bzk/flow/model/Entry.java
package net.bzk.flow.model;

import java.time.temporal.ChronoUnit;
import java.util.List;

import lombok.Data;
import lombok.EqualsAndHashCode;
import net.bzk.infrastructure.convert.OType;

@Data
public class Entry implements OType {

	private String clazz;
	private String boxUid;
	private boolean autoRegister;


	public Entry() {
		clazz = this.getClass().getName();
	}

	@Data
	@EqualsAndHashCode(callSuper = false)
	public static class FixedRateEntry extends Entry {

		private long period;
		private ChronoUnit unit;
		private long initialDelay;
		private String cronExpression;
		
	}
	
	@Data
	@EqualsAndHashCode(callSuper = false)
	public static class PluginEntry extends Entry{
		
		private List<String> requiredKeys;
		private List<String> outputKeys;
	}

	@Data
	@EqualsAndHashCode(callSuper = false)
	public static class EventEntry extends Entry{

		private Condition condition;

	}

}

File: flow-editor/src/app/model/entry.ts
import { VarService } from './../service/var.service';
import { ChronoUnit } from './enums';
import { OTypeClass } from '../utils/bzk-utils';
import { PropClazz, PropEnums, PropInfo, PropType } from '../utils/prop-utils';
import { OType } from './bzk-obj';


export class Entry extends OType {
  public boxUid: string;

  @PropInfo({
    title: 'auto Register',
    type: PropType.Boolean,
  })
  public autoRegister = false;
}

@OTypeClass({
  clazz: 'net.bzk.flow.model.Entry$FixedRateEntry'
})
@PropClazz({
  title: 'FixedRateEntry'
})
export class FixedRateEntry extends Entry {

  @PropInfo({
    title: 'period',
    type: PropType.Number
  })
  public period = 0;
  @PropInfo({
    title: 'unit',
    type: PropType.Enum
  })
  @PropEnums(Object.keys(ChronoUnit))
  public unit = ChronoUnit.HOURS;
  @PropInfo({
    title: 'initialDelay',
    type: PropType.Number
  })
  public initialDelay = 0;

}

@OTypeClass({
  clazz: 'net.bzk.flow.model.Entry$PluginEntry'
})
@PropClazz({
  title: 'PluginEntry'
})
export class PluginEntry extends Entry {

  @PropInfo({
    title: 'requiredKeys',
    type: PropType.List,
    child: {
      title: 'Key',
      type: PropType.Text,
      autocompleteFuncKey: 'ListAllVarKeys',
      newObj: 'TODO',
    },
  })
  public requiredKeys = new Array<string>();

  @PropInfo({
    title: 'outputKeys',
    type: PropType.List,
    child: {
      title: 'Key',
      type: PropType.Text,
      autocompleteFuncKey: 'ListAllVarKeys',
      newObj: 'TODO',
    },
  })
  public outputKeys = new Array<string>();

}


